File: benchmarks/tsb/bench_str_contains.ts
/**
 * Benchmark: str.contains() — regex and literal substring matching on 100k strings.
 * Outputs JSON: {"function": "str_contains", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series } from "../../src/index.ts";

const ROWS = 100_000;
const WARMUP = 5;
const ITERATIONS = 30;

const data = Array.from({ length: ROWS }, (_, i) => `item_${i % 500}_value_${i % 7}_end`);
const s = new Series({ data });

for (let i = 0; i < WARMUP; i++) {
  s.str.contains("value", false);
  s.str.contains("_[0-9]+_", true);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  s.str.contains("value", false);
  s.str.contains("_[0-9]+_", true);
  times.push(performance.now() - t0);
}

const total_ms = times.reduce((a, b) => a + b, 0);
console.log(
  JSON.stringify({
    function: "str_contains",
    mean_ms: Math.round((total_ms / ITERATIONS) * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(total_ms * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_str_contains.py
"""Benchmark: pd.Series.str.contains() — regex and literal substring matching on 100k strings."""
import json, time
import pandas as pd

ROWS = 100_000
WARMUP = 5
ITERATIONS = 30

data = [f"item_{i % 500}_value_{i % 7}_end" for i in range(ROWS)]
s = pd.Series(data)

for _ in range(WARMUP):
    s.str.contains("value", regex=False)
    s.str.contains(r"_[0-9]+_", regex=True)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    s.str.contains("value", regex=False)
    s.str.contains(r"_[0-9]+_", regex=True)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({
    "function": "str_contains",
    "mean_ms": round(total_ms / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total_ms, 3),
}))
